FlowCrypt/Functionality/DataManager/Encrypted Storage/KeyChainService.swift

```diff
@@ -19,67 +19,85 @@ actor KeyChainService {
-    @MainActor private func getKeychainIndex() throws -> String {
-        guard UIApplication.shared.isProtectedDataAvailable else {
-            // when not available, UserDefaults is empty and a new index would be wrongly generated
-            // which would cause new database encryption key to get generated
-            // and that would corrupt existing storage
-            // https://github.com/FlowCrypt/flowcrypt-ios/issues/1373
-            throw AppErr.general("KeyChainService: protected data is not available")
-        }
-        let dynamicPartIndex = "indexSecureKeychainPrefix"
-        if let storedDynamicPart = UserDefaults.standard.string(forKey: dynamicPartIndex) {
-            return constructKeychainIndex(dynamicPart: storedDynamicPart)
-        }
-        let newDynamicPart = try newRandomString()
-        UserDefaults.standard.set(newDynamicPart, forKey: dynamicPartIndex)
-        return constructKeychainIndex(dynamicPart: newDynamicPart)
+    private enum Constants {
+        static let legacyKeychainIndexPrefixInUserDefaults = "indexSecureKeychainPrefix"
+        static let keychainPropertyKey = "app-storage-encryption-key"
     }
 
-    @MainActor private func newRandomString() throws -> String {
-        logger.logInfo("newRandomString - generating new KeyChain index")
-        guard let randomBytes = CoreHost().getSecureRandomByteNumberArray(12) else {
-            throw AppErr.general("KeyChainService.newRandomString - randomBytes are nil")
+    @MainActor var storageEncryptionKey: Data {
+        get throws {
+            if let key = try fetchEncryptionKey() {
+                return key
+            }
+
+            // TODO: Should be removed after all users migrated to new keychain
+            if let legacyKey = try fetchLegacyEncryptionKey() {
+                try saveStorageEncryptionKey(data: legacyKey)
+                return legacyKey
+            }
+
+            return try generateStorageEncryptionKey()
         }
-        return Data(randomBytes)
-            .base64EncodedString()
-            .replacingOccurrences(of: "[^A-Za-z0-9]+", with: "", options: [.regularExpression])
     }
 
-    @MainActor private func constructKeychainIndex(dynamicPart: String) -> String {
-        return dynamicPart + "-indexStorageEncryptionKey"
+    @MainActor private func fetchLegacyEncryptionKey() throws -> Data? {
+        let userDefaults = UserDefaults.standard
+        let prefixKey = Constants.legacyKeychainIndexPrefixInUserDefaults
+
+        guard let storedDynamicPart = userDefaults.string(forKey: prefixKey) else { return nil }
+
+        userDefaults.removeObject(forKey: prefixKey)
+        let storageKey = storedDynamicPart + "-indexStorageEncryptionKey"
+
+        guard let encryptionKey = try fetchEncryptionKey(property: storageKey) else {
+            if try !EncryptedStorage.doesStorageFileExist {
+                throw AppErr.general("KeyChainService: failed to get encryption key prefix")
+            }
```

Comment:
This is odd - I think it should test the opposite:

```swift
        guard let encryptionKey = try fetchEncryptionKey(property: storageKey) else {
            // we got index in user defaults but could not retrieve encryption key from it
            if try EncryptedStorage.doesStorageFileExist {
                // the storage file was already initiated before - problem
                throw AppErr.general("KeyChainService: got legacy dynamic prefix from user defaults but could not find entry in key chain based on it")
            }
            // the storage was not initiated yet - no problem
            return nil
        }
```

Reply:
Ah, missed it when changing from `guard try` to `if try`, fixed now ✅